apps/ui/src/xchain/eth/EthWalletSigner.ts

```diff
@@ -44,60 +74,71 @@ export class EthWalletSigner extends AbstractWalletSigner<EthereumNetwork> {
-          ...Builder.WITNESS_ARGS.Secp256k1,
-          lock: signature,
-        }),
-      ),
-    ).serializeJson();
-    raw.witnesses = witnesses;
-    return raw as NervosNetwork['SignedTransaction'];
+  async approve(asset: EthereumNetwork['DerivedAssetIdent']): Promise<{ txId: string }> {
+    const erc20 = new ethers.Contract(asset, Erc20ABI, this.signer);
+    const transactionResponse = await erc20.approve(LockerContractAddress, ethers.constants.MaxUint256);
+    return { txId: transactionResponse.hash };
   }
-}
 
-function toNervosMessage(raw: NervosNetwork['RawTransaction'], witnesses: string[]): string {
-  const hasher = new Keccak256Hasher();
-  const txHash = new Blake2bHasher().hash(
-    new Reader(SerializeRawTransaction(normalizers.NormalizeRawTransaction(transformers.TransformRawTransaction(raw)))),
-  );
-  hasher.update(txHash);
-  const firstWitness = new Reader(witnesses[0]);
-  hasher.update(serializeBigInt(firstWitness.length()));
-  hasher.update(firstWitness);
-  for (let i = 1; i < witnesses.length; i++) {
-    const currentWitness = new Reader(witnesses[i]);
-    hasher.update(serializeBigInt(currentWitness.length()));
-    hasher.update(currentWitness);
+  async getAllowance(asset: EthereumNetwork['DerivedAssetIdent']): Promise<BigNumber> {
+    const erc20 = new ethers.Contract(asset, Erc20ABI, this.signer);
+    return erc20.allowance(await this.signer.getAddress(), LockerContractAddress);
   }
-  return hasher.digest().serializeJson();
-}
 
-function serializeBigInt(i: number) {
-  const view = new DataView(new ArrayBuffer(8));
-  view.setUint32(0, i, true);
-  return view.buffer;
+  async toPWTransaction(rawTx: NervosNetwork['RawTransaction']): Promise<Transaction> {
+    function toPWHashType(hashType: CKBComponents.ScriptHashType): HashType {
+      if (hashType === 'data') {
+        return HashType.data;
+      }
+      return HashType.type;
+    }
+
+    function toPWDepType(depType: CKBComponents.DepType): DepType {
+      if (depType === 'code') {
+        return DepType.code;
+      }
+      return DepType.depGroup;
+    }
+
+    const ckbRPC = new RPC(this._nervosRPCURL);
+    const inputs = await Promise.all(
+      rawTx.inputs.map((i: CKBComponents.CellInput) =>
+        Cell.loadFromBlockchain(ckbRPC, new OutPoint(i.previousOutput!.txHash, i.previousOutput!.index)),
+      ),
+    );
+    const outputs = rawTx.outputs.map(
+      (o, index) =>
+        new Cell(
+          new Amount(o.capacity, AmountUnit.shannon),
+          new Script(o.lock.codeHash, o.lock.args, toPWHashType(o.lock.hashType)),
+          o.type ? new Script(o.type.codeHash, o.type.args, toPWHashType(o.type.hashType)) : undefined,
+          undefined,
+          rawTx.outputsData[index],
+        ),
+    );
+    const cellDeps = rawTx.cellDeps.map(
+      (c) => new CellDep(toPWDepType(c.depType), new OutPoint(c.outPoint!.txHash, c.outPoint!.index)),
+    );
+    const pwLockCellDeps = [
```

Comment:
we should use the PW's constants [CHAIN_SPEC](https://github.com/lay2dev/pw-core/blob/861310b3dd8638f668db1a08d4c627db4c34d815/src/constants.ts#L13) rather than hard code here